contrib/dvn.py

```diff
@@ -0,0 +1,203 @@
+        self.mlflow_log_metric("prob random action", prob_random_action, step=episode_num)
+        if random.random() < prob_random_action:
+            action = self.env.action_space.sample()
+        else:
+            state_vals = tf.concat([self.v_model(np.asarray(self.env.get_afterstate(state, a)[0])[np.newaxis])
+                                for a in range(self.env.action_space.n)], 0)
+            action = tf.squeeze(tf.argmax(state_vals)).numpy()
+        return action
+
+    def _train(self):
+        train_acc_metric = keras.metrics.SparseCategoricalAccuracy()
+        game_scores = []
+        game_num_steps = []
+        for episode_num in range(self.params["num_episodes"]):
+            alpha = self.params["alpha0_ie_init_step_size"] / (1 + episode_num * self.params["alpha_decay"])
+            logging.debug(f"alpha: {alpha}")
+            self.mlflow_log_metric("alpha", alpha)
+            state = self.env.reset()
+            game_score = 0
+            for step_num in range(self.params["max_steps_per_episode"]):
+                action = self.get_action(episode_num, state)
+                next_state, reward, done, _ = self.env.step(action)
+                next_state = np.asarray(next_state)
+                self.memory.append((state,
+                                    action,
+                                    self.env.get_canonical_board(self.env.get_afterstate(state, action)[0]),
+                                    reward,
+                                    done,
+                                    next_state))
+                state = next_state
+                game_score += reward
+                if done:
+                    break
+
+            if episode_num >= self.params["learning_starts"]:
+                states, actions, afterstates, rewards, dones, next_states = self.memory.get_random_batch(self.params["batch_size"])
+                with tf.GradientTape() as tape:
+                    vals = self.v_model(afterstates)
+                    next_ca = []  # next_canonical_afterstates
+                    for n_a in range(self.env.action_space.n):  # n_a is next_action
+                        next_as = [self.env.get_afterstate(s, n_a)[0] for s in next_states]
+                        next_canonicals = np.asarray([self.env.get_canonical_board(s) for s in next_as]).reshape(next_states.shape)
+                        next_ca.append(self.v_model(next_canonicals))
+                    next_vals_all = tf.concat(next_ca, 1)
+                    next_vals = tf.expand_dims(tf.reduce_max(next_vals_all, axis=1), 1)
+                    disc_next_v = (1 - dones) * self.params["gamma_ie_discount_rate"] * next_vals
+                    td_target = rewards + disc_next_v
+                    val_targets = (1 - alpha) * vals + alpha * td_target
+                    loss = self.loss_fn(vals, val_targets)
```

Comment:
This is probably just me not understanding `GradientTape`, but how does the optimizer know which inputs to the NN (which, I believe, are canonical afterstates) are associated with which `vals`/`val_targets`?